clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/producer/AbstractKafkaBasedMessageSender.java

```diff
@@ -0,0 +1,331 @@
+            final Tracer tracer) {
+        Objects.requireNonNull(producerFactory);
+        Objects.requireNonNull(producerName);
+        Objects.requireNonNull(config);
+        Objects.requireNonNull(tracer);
+
+        this.producerFactory = producerFactory;
+        this.producerName = producerName;
+        this.config = config;
+        this.tracer = tracer;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Starts the producer.
+     */
+    @Override
+    public Future<Void> start() {
+        configureUniqueClientId();
+        getOrCreateProducer();
+        return Future.succeededFuture();
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Closes the producer.
+     */
+    @Override
+    public Future<Void> stop() {
+        return producerFactory.closeProducer(producerName);
+    }
+
+    /**
+     * Sends a message to a kafka cluster and doesn't wait for an outcome.
+     *
+     * @param topic The topic to send the message to.
+     * @param tenantId The tenant that the device belongs to.
+     * @param deviceId The device identifier.
+     * @param payload The data to send.
+     * @param properties Additional meta data that should be included in the message.
+     * @param context The currently active OpenTracing span (may be {@code null}). An implementation should use this as
+     *            the parent for any span it creates for tracing the execution of this operation.
+     * @throws NullPointerException if topic, tenantId, deviceId or properties are {@code null}.
+     */
+    protected void send(final String topic, final String tenantId, final String deviceId, final Buffer payload,
+            final Map<String, Object> properties, final SpanContext context) {
+
+        Objects.requireNonNull(topic);
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+        Objects.requireNonNull(properties);
+
+        final Span span = startSpan(topic, tenantId, deviceId, References.FOLLOWS_FROM, context);
+        final List<KafkaHeader> headers = encodePropertiesAsKafkaHeaders(properties, span);
+        send(topic, tenantId, deviceId, payload, headers, span);
+    }
+
+    /**
+     * Sends a message to a kafka cluster and doesn't wait for an outcome.
+     *
+     * @param topic The topic to send the message to.
+     * @param tenantId The tenant that the device belongs to.
+     * @param deviceId The device identifier.
+     * @param payload The data to send.
+     * @param headers Additional meta data that should be included in the message.
+     * @param context The currently active OpenTracing span (may be {@code null}). An implementation should use this as
+     *            the parent for any span it creates for tracing the execution of this operation.
+     * @throws NullPointerException if topic, tenantId, deviceId or headers are {@code null}.
+     */
+    protected void send(final String topic, final String tenantId, final String deviceId, final Buffer payload,
```

Comment:
I was first wondering whether we should better use a `Future<Void>` return value here - similar to corresponding AMQP `AbstractSender` method. That would allow us to return a failed Future if the process of sending the message itself has failed (e.g. because the producer is already closed or there was a serialization error).

Unfortunately, the Vert.x Kafka client doesn't provide this feature, to distinguish between such a producer side error and an error related to the asynchronous acknowledgement.
So, for the time being, we have to accept not being able to get these early exceptions here and can only define a `void` send method here.

Reply:
Initially I had the same thought of returning `Future<Void>` and I didn't see any advantage in case of Vertx Kafka clients and hence changed to `void`.